File: OrangeJuice.Server.Core/Validation/XmlRequestValidator.cs
using System.Xml;
using System.Xml.Linq;
using System.Xml.XPath;

namespace OrangeJuice.Server.Validation
{
	public sealed class XmlRequestValidator : IValidator<XElement>
	{
		public bool IsValid(XElement item)
		{
			if (item == null)
				return false;

			XmlNamespaceManager nm = new XmlNamespaceManager(new NameTable());
			nm.AddNamespace("x", item.Name.Namespace.ToString());

			return (bool)item.XPathSelectElement("x:Request/x:IsValid", nm);
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Services/XmlRequestValidator.fs
﻿namespace OrangeJuice.Server.FSharp.Validation

open System.Xml
open System.Xml.Linq
open System.Xml.XPath

open OrangeJuice.Server

type XmlRequestValidator() =
    interface IValidator<XElement> with
        member this.IsValid(item : XElement) : bool =
            if item = null then false
            else
                let nm = new XmlNamespaceManager(new NameTable())
                nm.AddNamespace("x", item.Name.Namespace.ToString()) |> ignore

                let toBool e =
                    match e with
                    | null -> None
                    | e -> Some(XElement.op_Explicit e : bool)

                (item.XPathSelectElement("x:Request/x:IsValid", nm) |> toBool).Value
